proposals/0000-componentized-cabal.rst

```diff
@@ -0,0 +1,375 @@
+   various files the component may install, we will apply different
+   defaults to these paths::
+
+        Variable    Current         New
+        -----------------------------------------
+        bindir      $prefix/bin     (same)
+        libsubdir   $abi/$libname   $abi/$cid
+        datasubdir  $abi/$pkgid     $abi/$cid
+        docdir      $datadir/doc/$abi/$pkgid     $datadir/doc/$abi/$cid
+
+6. As you might expect, subsequent ``build``, ``register``, etc commands
+   operate only on the enabled component.
+
+There are a few expectations as to how this per-component configure
+interface is intended to be used by a client:
+
+1. If the client reuses the same source directory to build multiple
+   components, it must assign a distinct build directory ``--distdir`` to
+   each component build.
+
+2. The installation directories configured for each component are
+   expected to be disjoint for each component. This is encouraged by the
+   different default install directories. 
+
+3. A component must be copied and registered to a package database
+   within the package database, before any of its (same-package)
+   dependents can be built.
+
+4. The ordering components are built should respect internal
+   ``build-tools`` dependencies.
+
+**Interaction with hooks.** This proposal leaves the hooks API
+unchanged. Thus, just as before, hooks are responsible for determining
+what components are enabled before performing the operations they need.
+If they operate on the available ``ComponentLocalBuildInfo`` hooks should
+automatically work properly with these changes (even better than
+manually specifying which components to build at ``./Setup build`` time)
+
+**Interaction with Haddock.** Initially, ``haddock`` will be kept as it is
+today: a command which builds Haddock documentation for every enabled
+component (one in this case). However, eventually, it would be good to
+treat building Haddock documentation as a "component" in and of itself,
+so it can be generated and installed independently of building any
+component (indeed, Haddock is completely independent of a build).
+Thus, every component would also have a corresponding Haddock component
+for the documentation.
+
+The primary complication is that cross-linking for Haddock documentation
+is implemented by storing paths for haddock interfaces and HTMLs in
+the package database
+(``hadock-interfaces`` and ``haddock-html``).  Clearly, since we want
+to build Haddocks independently of libraries, this should no longer
+be done.  I suggest we workaround this problem by making "fake"
+installed package database entries to represent Haddock documentation:
+Haddock components form an independent, parallel dependency tree
+to the library tree.
+
+**Interaction with package common files.** Cabal packages can be bundled
+with package-common files (specified by the ``data-files`` field), which
+are installed to a location common over all components inside a package.
```

Comment:
No. `dataFiles` is property of [`PackageDescription`](http://hackage.haskell.org/package/Cabal-1.24.0.0/docs/Distribution-PackageDescription.html#t:PackageDescription)

But that's relatively easy to change (in a backward compatible way)